### ISSUE TITLE
Add an `.element` method to the `Ui` that returns the entire `Element` for the GUI (as an alternative to `.draw`).

### ISSUE THREAD

**Issue by mitchmindtree** [PistonDevelopers/conrod#527]
> This would allow a user to use their own renderer and to position the gui using elmesque layout, etc.

As mentioned in #439

**Comment by jarrett** [comment]
I could attempt this.

**Comment by jarrett** [comment]
I guess `Graph` will need an `element` method, which `Ui::element` will call. I will probably need to split out some of the code in `Graph::draw` so it can be shared between `draw` and `element.`

I see the at `Graph::draw` iterates over the graph nodes, [requesting an `Element` from each one](https://github.com/PistonDevelopers/conrod/blob/master/src/graph/mod.rs#L416). I'm thinking of doing something similar with `collage`. Obviously drawing order is very important. Does `collage` guarantee that order of the `Vec<Form>` corresponds to the drawing order? I.e. is the tail of the `Vec` guaranteed to be drawn last?

**Comment by mitchmindtree** [comment]
Cool! And yep, that sounds great. Yes the `Form`s in the collage `Vec` are always drawn in order.

I'd recommend maybe using the [`layers` function](https://github.com/mitchmindtree/elmesque/blob/master/src/element.rs#L335) instead - that way you won't have to convert the `Element`s back to `Form`s to be able to use them in `collage`. I *think* that should work (not sure `layers` has been tested)?

**Comment by mitchmindtree** [comment]
I'm imagining something along these lines:
```Rust
let elements = depth_order.iter().filter_map(|&idx| {
    if let &mut Node::Widget(ref mut container) = &mut graph[idx] {
        if container.has_updated {
            container.has_updated = false;
            return Some(container.element.clone());
        }
    }
    None
});
layers(elements)
```

**Comment by jarrett** [comment]
I've [gotten started on this](https://github.com/jarrett/conrod/commit/2c19b006e7aab57768bf68c7cfb383942b19f638). I'm moving towards writing `Ui::element` now.

Right now the `Ui` owns the glyph cache. We probably don't want people creating one glyph cache for Conrod and another for general Elmesque drawing. In my Elmesque example, I avoided that by [borrowing the cache from the `UI`](https://github.com/PistonDevelopers/conrod/blob/c1a81e3ee53f990b61c9b9a1af9788c8f4049011/examples/elmesque.rs#L79) when constructing a `Renderer`.

Should we change that somehow? In applications that call `Ui::element` instead of `Ui::draw`, the `Ui` will never need the glyph cache to render. It will [need the glyph cache to update](https://github.com/PistonDevelopers/conrod/blob/c1a81e3ee53f990b61c9b9a1af9788c8f4049011/src/widget/mod.rs#L362), though. So I wonder if there's a way we can get a glyph cache reference to the widget on update without having the `UI` own the glyph cache.

**Comment by jarrett** [comment]
I've added `Ui::element`. [Here's the branch](https://github.com/PistonDevelopers/conrod/compare/master...jarrett:elmesque-integration). Does this look like the API you had in mind?

**Comment by mitchmindtree** [comment]
@jarrett that branch looks great! Will be nice to have this :+1:

Re glyph cache: I agree that we could handle this better - For example, it would be nice if we could also offer the option of sharing the glyph cache (taking a `Rc<GlyphCache>` or `Arc<GlyphCache>` rather than an owned one), however I'm not sure how we can expose this nicely just yet. I've heard that the possible future addition of HKT may be useful for handling this situation nicely, but I think that's a while off. Either way, we can probably deal with this later in another issue :)

**Comment by mitchmindtree** [comment]
Closed by #538
